lib/src/package_config_impl.dart

```diff
@@ -154,45 +169,70 @@ class SimplePackage implements Package {
   /// Creates a [SimplePackage] with the provided content.
   ///
   /// The provided arguments must be valid.
-  factory SimplePackage(String name, Uri root, Uri packageUriRoot,
-      String /*?*/ languageVersion, dynamic extraData) {
-    _validatePackageData(name, root, packageUriRoot, languageVersion);
+  ///
+  /// If the arguments are invalid then the error is reported by
+  /// calling [onError], then the erroneous entry is ignored.
+  ///
+  /// Returns `null` if the input is invalid and an approximately valid package
+  /// cannot be salvaged from the input.
+  static SimplePackage /*?*/ validate(
+      String name,
+      Uri root,
+      Uri packageUriRoot,
+      String /*?*/ languageVersion,
+      dynamic extraData,
+      void onError(Object error)) {
+    bool fatalError = false;
+    if (!isValidPackageName(name)) {
+      onError(
+          PackageConfigArgumentError(name, "name", "Not a valid package name"));
+    }
+    if (root.isScheme("package")) {
+      onError(PackageConfigArgumentError("$root", "root",
+          "Must not be a package URI"));
+      fatalError = true;
+    } else if (!isAbsoluteDirectoryUri(root)) {
+      onError(PackageConfigArgumentError(
+          "$root",
+          "root",
+          "Not an absolute URI with no query or fragment "
+              "with a path ending in /"));
+      // Try to recover. If the URI has a scheme,
+      // then ensure that the path ends with `/`.
+      if (!root.hasScheme) {
+        fatalError = true;
+      } else if (!root.path.endsWith("/")) {
+        root = root.replace(path: root.path + "/");
+      }
+    }
+    if (!fatalError) {
+      if (!isAbsoluteDirectoryUri(packageUriRoot)) {
+        onError(PackageConfigArgumentError(
+            packageUriRoot,
+            "packageUriRoot",
+            "Not an absolute URI with no query or fragment "
+                "with a path ending in /"));
+        packageUriRoot = root;
+      } else if (!isUriPrefix(root, packageUriRoot)) {
+        onError(PackageConfigArgumentError(packageUriRoot, "packageUriRoot",
+            "The package URI root is not below the package root"));
+        packageUriRoot = root;
+      }
+    }
+    if (languageVersion != null &&
+        checkValidVersionNumber(languageVersion) >= 0) {
+      onError(PackageConfigArgumentError(languageVersion, "languageVersion",
+          "Invalid language version format"));
+      languageVersion = null;
```

Comment:
Could we get a special marker for invalid language version?

Reply:
What kind of "special marker"?

Reply:
So a client can tell that a language version was set, although it was invalid. It my suggested hack-patch I simply overwrite the string "invalid" instead of null --- though something more elegant is surely possible...